packages/runtime/src/runtimes/react/components/Element.tsx

```diff
@@ -1,66 +1,72 @@
 import { ElementData } from './ElementData'
+import { ElementImperativeHandle } from '../element-imperative-handle'
 import { FindDomNode } from '../find-dom-node'
 
-type ElementProps = {
-  element: ElementDataOrRef
+type ElementWithFallbackProps = {
+  element: ReactPageElement
+  fallback?: ReactNode
+  isInitialData?: boolean
 }
 
-export const Element = memo(
+export const ElementWithFallback = memo(
   forwardRef(function Element(
-    { element }: ElementProps,
+    { element, fallback, isInitialData = false }: ElementWithFallbackProps,
     ref: Ref<ElementImperativeHandle>,
-  ): JSX.Element {
-    const elementKey = element.key
-    const dispatch = useDispatch()
-    const documentKey = useDocumentKey()
+  ): JSX.Element | null {
+    const showFallback = fallback != null && isInitialData
+
     const useFindDomNodeRef = useRef(true)
     const imperativeHandleRef = useRef(new ElementImperativeHandle())
+
     const findDomNodeCallbackRef = useCallback((current: (() => Element | Text | null) | null) => {
       if (useFindDomNodeRef.current === true) {
         imperativeHandleRef.current.callback(() => current?.() ?? null)
       }
     }, [])
+
     const elementCallbackRef = useCallback((current: unknown | null) => {
       useFindDomNodeRef.current = false
 
       imperativeHandleRef.current.callback(() => current)
     }, [])
-    const isRegisterElementDisabled = useDisableRegisterElement()
 
     useImperativeHandle(ref, () => imperativeHandleRef.current, [])
 
-    useEffect(() => {
-      if (documentKey == null || isRegisterElementDisabled) return
-
-      return dispatch(
-        registerComponentHandleEffect(documentKey, elementKey, imperativeHandleRef.current),
-      )
-    }, [dispatch, documentKey, elementKey, isRegisterElementDisabled])
-
-    useEffect(() => {
-      if (documentKey == null || isRegisterElementDisabled) return
-
-      return dispatch(mountComponentEffect(documentKey, elementKey))
-    }, [dispatch, documentKey, elementKey, isRegisterElementDisabled])
-
     return (
       <FindDomNode ref={findDomNodeCallbackRef}>
-        {isElementReference(element) ? (
-          <ElementReference key={elementKey} ref={elementCallbackRef} elementReference={element} />
-        ) : (
-          <ElementData key={elementKey} ref={elementCallbackRef} elementData={element} />
-        )}
+        <ElementRegistration componentHandle={imperativeHandleRef.current} elementKey={element.key}>
+          {showFallback ? (
+            fallback
+          ) : isElementReference(element) ? (
+            <ElementReference
+              key={element.key}
+              ref={elementCallbackRef}
+              elementReference={element}
+            />
+          ) : (
+            <ElementData key={element.key} ref={elementCallbackRef} elementData={element} />
+          )}
+        </ElementRegistration>
       </FindDomNode>
     )
   }),
 )
+
+type ElementProps = {
+  element: ReactPageElement
```

Comment:
Seems this changed from `ElementDataOrRef` to `ReactPageElement`. How come?

Reply:
This is a mistake. The underlying type is actually the same, it was just being imported under a different alias. Must have gotten switched by mistake during all the shuffling. Will address.